### ISSUE TITLE
ros-melodic/tf and others depends on boost_signals and cant emerge

### ISSUE THREAD

**Issue by fcolecumberri** [ros/ros-overlay#971]
as described https://github.com/ros/ros-overlay/issues/912 some packages needs boost signals which are no longer available, the fixes described there works with ros-kinetic, but the same problem exist for ros-melodic.

**Comment by awesomebytes** [comment]
Hello @fcolecumberri I've spent some time lately either pushing fixes upstream or asking for re-releases to fix this issue. As far as I know it has been fixed for most packages I found. Including ros-melodic/tf.
https://github.com/ros/geometry/commit/2eed72cb7b63fc4ee98af695c625d9a741a09897

And the ebuild reflects this version: https://github.com/ros/ros-overlay/tree/master/ros-melodic/tf

You may need to sync your overlay?

**Comment by fcolecumberri** [comment]
first of all, sorry, I didn't mean tf, ment tf2, but I still get the error

~~~
Could not find the following Boost libraries:

    boost_signals
~~~

now that I have digged a little bit  more, I think it is already solved in version 0.6.6-1 and later, but as you can see in https://github.com/ros/ros-overlay/tree/master/ros-melodic/tf2 the current version in the overlay is the 6.5 so the problem is still there.

thanks a lot.

**Comment by awesomebytes** [comment]
That's strange. I swear I ran a superflore command with tf2 in it 2 days ago (and the last release was 7 days ago) and I got no changes... Hmm.

Could you do (installing superflore and configuring it):

`superflore-gen-ebuilds --ros-distro melodic --only tf2`

It should generate a PR with the ebuild updated. And if it doesn't work we'll need to investigate why.

I'm away of a computer for a bit, so I can't do it myself.

**Comment by fcolecumberri** [comment]
I'm almost sure that the reason is that the ebuild ( https://github.com/ros/ros-overlay/blob/master/ros-melodic/tf2/tf2-0.6.5.ebuild ) is downloading from https://github.com/ros-gbp/geometry2-release/archive/release/melodic/tf2/0.6.5-0.tar.gz which is an old version, instead of https://github.com/ros-gbp/geometry2-release/archive/release/melodic/tf2/0.6.6-1.tar.gz 

I am installing the superflore to make the stuff you asked.

**Comment by fcolecumberri** [comment]
Ok, I tried to run the superflore-gen-ebuilds but I got:
~~~
Traceback (most recent call last):
  File "/usr/lib64/python3.6/site-packages/pkg_resources/__init__.py", line 583, in _build_master
    ws.require(__requires__)
  File "/usr/lib64/python3.6/site-packages/pkg_resources/__init__.py", line 900, in require
    needed = self.resolve(parse_requirements(requirements))
  File "/usr/lib64/python3.6/site-packages/pkg_resources/__init__.py", line 791, in resolve
    raise VersionConflict(dist, req).with_context(dependent_req)
pkg_resources.ContextualVersionConflict: (rosdep 0.15.0 (/usr/lib64/python3.6/site-packages), Requirement.parse('rosdep>=0.15.2'), {'superflore'})

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python-exec/python3.6/superflore-gen-ebuilds", line 6, in <module>
    from pkg_resources import load_entry_point
  File "/usr/lib64/python3.6/site-packages/pkg_resources/__init__.py", line 3251, in <module>
    @_call_aside
  File "/usr/lib64/python3.6/site-packages/pkg_resources/__init__.py", line 3235, in _call_aside
    f(*args, **kwargs)
  File "/usr/lib64/python3.6/site-packages/pkg_resources/__init__.py", line 3264, in _initialize_master_working_set
    working_set = WorkingSet._build_master()
  File "/usr/lib64/python3.6/site-packages/pkg_resources/__init__.py", line 585, in _build_master
    return cls._build_from_requirements(__requires__)
  File "/usr/lib64/python3.6/site-packages/pkg_resources/__init__.py", line 598, in _build_from_requirements
    dists = ws.resolve(reqs, Environment())
  File "/usr/lib64/python3.6/site-packages/pkg_resources/__init__.py", line 786, in resolve
    raise DistributionNotFound(req, requirers)
pkg_resources.DistributionNotFound: The 'rosdep>=0.15.2' distribution was not found and is required by superflore
~~~
since it needs a version of rosdep greater than the current one in the overlay.

**Comment by awesomebytes** [comment]
The error comes from your local system, you need to update your local
rosdep... Which if you are using Gentoo as your main system may be a bit
problematic? Or maybe not using pip to update it.

@allenh1 may want to provide some feedback on that?

You also want to run `rosdep update` before doing that so the rosdep keys
are updated.

**Comment by fcolecumberri** [comment]
ok, I have moreless manae it to rn, but now I get:
~~~
Traceback (most recent call last):
  File "/usr/bin/superflore-gen-ebuilds", line 11, in <module>
    load_entry_point('superflore==0.3.1+4.ga008d82', 'console_scripts', 'superflore-gen-ebuilds')()
  File "/usr/lib64/python3.6/site-packages/superflore-0.3.1+4.ga008d82-py3.6.egg/superflore/generators/ebuild/run.py", line 50, in main
superflore.exceptions.NoGitHubAuthToken
~~~
even while I have the SUPERFLORE_GITHUB_TOKEN set a personal access tokens

**Comment by fcolecumberri** [comment]
ok, I manage to make the auth work (and lots of docker issues related to the fact that I have bearly used docker in my life), but now I get:
~~~
>>>> "melodic" distro detected...
>>>> Working in temporary directory /tmp/tmp68t1yz2x
>>>> Creating new branch gentoo-bot-DsvcrHCCmk...
>>>> 
>>>> Regenerating package 'tf2'...
>>>> Cleaning up temporary directory /tmp/tmp68t1yz2x
Traceback (most recent call last):
  File "/usr/bin/superflore-gen-ebuilds", line 11, in <module>
    load_entry_point('superflore==0.3.1+4.ga008d82', 'console_scripts', 'superflore-gen-ebuilds')()
  File "/usr/lib64/python3.6/site-packages/superflore-0.3.1+4.ga008d82-py3.6.egg/superflore/generators/ebuild/run.py", line 133, in main
  File "/usr/lib64/python3.6/site-packages/superflore-0.3.1+4.ga008d82-py3.6.egg/superflore/generators/ebuild/gen_packages.py", line 86, in regenerate_pkg
  File "/usr/lib64/python3.6/site-packages/superflore-0.3.1+4.ga008d82-py3.6.egg/superflore/generators/ebuild/gen_packages.py", line 217, in ebuild_text
  File "/usr/lib64/python3.6/site-packages/superflore-0.3.1+4.ga008d82-py3.6.egg/superflore/generators/ebuild/ebuild.py", line 196, in get_ebuild_text
  File "/usr/lib64/python3.6/site-packages/superflore-0.3.1+4.ga008d82-py3.6.egg/superflore/utils.py", line 215, in resolve_dep
  File "/usr/lib64/python3.6/site-packages/superflore-0.3.1+4.ga008d82-py3.6.egg/superflore/rosdep_support.py", line 95, in resolve_rosdep_key
  File "/usr/lib64/python3.6/site-packages/superflore-0.3.1+4.ga008d82-py3.6.egg/superflore/rosdep_support.py", line 51, in get_view
  File "/usr/lib64/python3.6/site-packages/rosdep-0.18.0-py3.6.egg/rosdep2/catkin_support.py", line 119, in get_catkin_view
    """ % (sources_list_dir))
rosdep2.catkin_support.ValidationFailed: rosdep database does not have any sources.
Please make sure you have a valid configuration in:
	/etc/ros/rosdep/sources.list.d
~~~
even when I have /etc/ros/rosdep/sources.list.d/20-default.list generated by rosdep

**Comment by awesomebytes** [comment]
I'm sorry you are running into so many problems. I'm not that experienced
with superflore. I use it from Ubuntu 18.04 and there I had no issues
(installing from source tho, to have the last version).

@allenh1 should have more feedback I hope.

I'll do that command when I'm back on a computer later tonight (Sydney
time).

**Comment by awesomebytes** [comment]
I tried to run superflore myself and something is wrong...
```bash
superflore-gen-ebuilds --ros-distro melodic --only tf2
>>>> "melodic" distro detected...
>>>> Working in temporary directory /tmp/tmpt8f4_bvv
>>>> Creating new branch gentoo-bot-ZklajPnszh...
>>>> 
>>>> Regenerating package 'tf2'...
>>>> Successfully generated ebuild for package 'tf2'.
>>>> Pulling docker image 'allenh1/ros_gentoo_base:latest'...
>>>> Running docker image...
>>>> Generating manifests...
>>>> Working in temporary directory /tmp/tmpzg9el3p3
>>>> Running container with command string 'bash -c 'cd /tmp/ros-overlay/ros-melodic/tf2 &>> /tmp/tmpzg9el3p3/log.txt && repoman manifest &>> /tmp/tmpzg9el3p3/log.txt''...
>>>> Docker container exited.
>>>> Cleaning up temporary directory /tmp/tmpzg9el3p3
>>>> Adding changes...
>>>> Nothing changed; no commit done
>>>> Forking repository if a fork does not exist...
>>>> Pushing changes to fork...
>>>> Filing pull-request...
!!!! Failed to file PR with the ros/ros-overlay repo!
!!!! Exception: 422 {"message": "Validation Failed", "errors": [{"resource": "PullRequest", "code": "custom", "message": "No commits between ros:master and awesomebytes:gentoo-bot-ZklajPnszh"}], "documentation_url": "https://developer.github.com/v3/pulls/#create-a-pull-request"}
>>>> Cleaning up temporary directory /tmp/tmpt8f4_bvv
```

@allenh1 there is a new version for sure...

**Comment by allenh1** [comment]
@awesomebytes Just did a vanilla superflore command, seems fine (see #972 ).

I think that's just a poorly-worded way of saying "no changes were made". I didn't see an update for `tf2` in that PR either...

**Comment by allenh1** [comment]
@awesomebytes it is likely pulled in by a different package. I'm going to completely regenerate `melodic`, that should clear most of this up.

**Comment by flabrosse** [comment]
I confirm that tf2 is still looking for boost_signals.

**Comment by awesomebytes** [comment]
@allenh1 tf2 HAS a new version: hhttps://github.com/ros-gbp/geometry2-release#geometry2-melodic---066-1 0.6.6-1

Meanwhile we have 0.6.5 in ros-overlay: https://github.com/ros/ros-overlay/tree/master/ros-melodic/tf2

And superflore does say there isn't any change. When there is. Something is fishy.

**Comment by allenh1** [comment]
> @allenh1 tf2 HAS a new version: https://github.com/ros-gbp/geometry2-release#geometry2-melodic---066-1 0.6.6-1

@awesomebytes as of right now, it looks like the release has not been made into [`rosdistro/distribution.yaml`](https://github.com/ros/rosdistro/blob/a6d8a071c4701a2c18e6a25b81429b056390f90b/melodic/distribution.yaml#L3086-L3113).

Superflore won't pick up the change until it is modified in that file.

**Comment by awesomebytes** [comment]
Thanks for the clarification @allenh1 . I didn't know that :) and trying to dig into the code didn't get me to there, I guess I didn't dig deep enough!

I see there is a PR pending on rosdistro. I'll wait for that.

**Comment by flabrosse** [comment]
Today there was a new version of ROS melodic, so I was hoping for a fix.  tf2 still does not merge because of boost signal.  It seems that what is holding the PR mentioned above is the rviz Manifest conflicting.  How is that related to the tf2 problem?

**Comment by awesomebytes** [comment]
@flabrosse Apparently upstream they havent updated tf2 yet: https://github.com/ros/rosdistro/blob/master/melodic/distribution.yaml#L3115-L3136 :/

If https://github.com/ros/ros-overlay/pull/978 is fixed to be able to merge, would be good to merge it. @allenh1  any input?

**Comment by fcolecumberri** [comment]
@flabrosse create a localrepo (https://wiki.gentoo.org/wiki/Handbook:AMD64/Portage/CustomTree#Defining_a_custom_repository https://wiki.gentoo.org/wiki/Custom_repository) and try the tf2 file from #978  

here it is:
[tf2-0.6.6.ebuild.tar.gz](https://github.com/ros/ros-overlay/files/4483754/tf2-0.6.6.ebuild.tar.gz)
it worked for me

**Comment by allenh1** [comment]
> If #978 is fixed to be able to merge, would be good to merge it. @allenh1 any input?

Yeah, looks fine to me. @fcolecumberri I appreciate the patches!

@flabrosse Sorry for all this. I've been away from my Gentoo boxes as of late, but just fell in love with them again (and am on mine right now, actually). Thanks you for testing all this! As bad as I feel when I get bug reports, I feel worse when I don't (cause at least I know people are using it when I get word it's broken).

Cheers!

**Comment by flabrosse** [comment]
@allenh1 No worries, and thank you for the hard work on this (ROS on gentoo).  I appreciate it a lot.

**Comment by flabrosse** [comment]
Another one is looking for signals: hector_mapping.  I guess it's the same fix?

**Comment by allenh1** [comment]
@flabrosse I think so, yeah. Likely the same fix on all of them. Which is kind of obnoxious.

**Comment by flabrosse** [comment]
To pick up the comment I wrongly made in another issue, tf now emerges fine, but not hector_mapping.  Still can't find boost signals.

**Comment by awesomebytes** [comment]
I just did a PR: https://github.com/tu-darmstadt-ros-pkg/hector_slam/pull/73

To patch it. But it will take a while to get through for sure (merging, re-release, update of ros-overlay...).

Meanwhile you can use this patch: https://gist.github.com/awesomebytes/d5ef49533165d698d3bf2fcc0e630bd2

Usable as:
```bash
mkdir -p $EPREFIX/etc/portage/patches/ros-melodic/hector_mapping
cd $EPREFIX/etc/portage/patches/ros-melodic/hector_mapping
wget https://gist.githubusercontent.com/awesomebytes/d5ef49533165d698d3bf2fcc0e630bd2/raw/bda054024b15f4451281c9f8cbea63bc2fef503b/0001-Remove-signals-to-enable-build-with-boost-1.69.patch
```

**Comment by flabrosse** [comment]
Hector_mapping now emerges!

Thank you to all involved!

**Comment by awesomebytes** [comment]
Great! Let's close this :)
